[mlir][tensor] Extend the logic to generalise tensor.pack

banach-space opened the pull request:
Extends the logic to generalise tensor.pack (into e.g. tensor.pad +
tensor.transpose) so that it also works when one of the inner tile sizes
is scalable (i.e. a multiple of `vector.vscale`). For example:
```mlir
  %c8 = arith.constant 8 : index
  %vscale = vector.vscale
  %c8_vscale = arith.muli %vscale, %c8 : index
  %0 = tensor.pack %input
      padding_value(%pad : f32)
      inner_dims_pos = [0, 1]
      inner_tiles = [%c8_vscale, 2]
      into %output : tensor<5x1xf32> -> tensor<1x1x?x2xf32>
}
```
is generalised as:
```mlir
  %c8 = arith.constant 8 : index
  %vscale = vector.vscale
  %c8_vscale = arith.muli %vscale, %c8 : index
  %0 = affine.apply #map()[%c8_vscale, %c5]
  %padded = tensor.pad %arg0 low[0, 0] high[%0, 1] {
  ^bb0(%arg3: index, %arg4: index):
    tensor.yield %arg2 : f32
  } : tensor<5x1xf32> to tensor<?x2xf32>
```

At the Tensor level, we model scalability using dynamic shapes and this
change basically extends the relevant logic so that it also works for
dynamic shapes.

banach-space committed:
[mlir][tensor] Extend the logic to generalise tensor.pack

Extends the logic to generalise tensor.pack (into e.g. tensor.pad +
tensor.transpose) so that it also works when one of the inner tile sizes
is scalable (i.e. a multiple of `vector.vscale`). For example:
```mlir
  %c8 = arith.constant 8 : index
  %vscale = vector.vscale
  %c8_vscale = arith.muli %vscale, %c8 : index
  %0 = tensor.pack %input
      padding_value(%pad : f32)
      inner_dims_pos = [0, 1]
      inner_tiles = [%c8_vscale, 2]
      into %output : tensor<5x1xf32> -> tensor<1x1x?x2xf32>
}
```
is generalised as:
```mlir
  %c8 = arith.constant 8 : index
  %vscale = vector.vscale
  %c8_vscale = arith.muli %vscale, %c8 : index
  %0 = affine.apply #map()[%c8_vscale, %c5]
  %padded = tensor.pad %arg0 low[0, 0] high[%0, 1] {
  ^bb0(%arg3: index, %arg4: index):
    tensor.yield %arg2 : f32
  } : tensor<5x1xf32> to tensor<?x2xf32>
```

At the Tensor level, we model scalability using dynamic shapes and this
change basically extends the relevant logic so that it also works for
dynamic shapes. However, rather than allowing arbitrary values and
number of tile sizes to be dynamic, only _one_ tile size is allowed to
be dynamic. In addition, it is required to be a constant multiple of
`vector.vscale`.

While the requirements above can be relaxed, I wanted to avoid full
generality for now. Primarily to avoid complexity that's not yet needed
and to make reviewing a bit easier.

hanhanW commented on mlir/include/mlir/Dialect/Tensor/Utils/Utils.h:
```diff
@@ -14,12 +14,13 @@
-// (i.e. it adds trailing padding values until the desired size is met).
+// Return a PadOp that pads `source` to `type` size. Output sizes (from `type`)
+// are assumed to be static and greater than the potentially dynamic input sizes
+// (from `source). The op performs "high" padding (i.e. it adds trailing padding
```
```suggestion
// (from `source`). The op performs "high" padding (i.e. it adds trailing padding
```

hanhanW commented on mlir/lib/Dialect/Tensor/Utils/Utils.cpp:
```diff
@@ -16,26 +16,48 @@
-                                    OpBuilder &b) {
+                                    OpBuilder &b,
+                                    std::optional<Value> dynOutDim) {
+  assert(llvm::count_if(
+             type.getShape(),
+             [](int64_t dim) { return ShapedType::isDynamic(dim); }) <= 1 &&
+         "At most one output dim can be dynamic!");
```
nit: I think we can use `type.getNumDynamicDims() <= 1` in the assertion.

hanhanW commented on mlir/lib/Dialect/Tensor/Utils/Utils.cpp:
```diff
@@ -16,26 +16,48 @@
-    AffineExpr d0;
-    bindDims(b.getContext(), d0);
-    OpFoldResult sz = tensor::getMixedSize(b, loc, source, en.index());
-    high[en.index()] =
-        affine::makeComposedFoldedAffineApply(b, loc, en.value() - d0, {sz});
+    if (!ShapedType::isDynamic(en.value())) {
```
As we're touching the code, can we switch to structured binding declaration? E.g.,

```
for (const auto [idx, val] : enumerate(type.getShape()))
```

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
```diff
@@ -1021,8 +1021,16 @@ LogicalResult ExtractSliceOfPadTensorSwapPattern::matchAndRewrite(
+///
+/// This method assumes that all outer dims for this pack Op are 1.
+///
+/// At most _one_ inner tile size can be _dynamic_, all other inner tiles are
+/// required to have static sizes. The inner tile that's dynamic must be a
+/// multiple of vector.vscale (to support scalable tile sizes). This condition
+/// can be relaxed in the future.
```
My motivation was to start conservatively and then generalise when needed. But I agree, we should leak concepts such as "scalability" to this layer.

banach-space commented on mlir/lib/Dialect/Tensor/Utils/Utils.cpp:
```diff
@@ -16,26 +16,48 @@
+      // output dims. Compute the padding width.
+      AffineExpr d0;
+      bindDims(b.getContext(), d0);
+      OpFoldResult sz = tensor::getMixedSize(b, loc, source, en.index());
```
Thanks for the suggestion! This gave me an idea for even more re-use. Let me extend what you are proposing here :)

banach-space committed:
fixup! [mlir][tensor] Extend the logic to generalise tensor.pack

Address PR comments from Han-Chung. Some clean-up and also relaxing the
requirement that the dynamic dim has to be a constant multiple of
vector.vscale.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
```diff
@@ -1181,7 +1221,23 @@ LogicalResult GeneralizeOuterUnitDimsPackOpPattern::matchAndRewrite(
   SmallVector<int64_t> transpShape = readShape;
   applyPermutationToVector<int64_t>(transpShape, perm);
 
-  Value empty = rewriter.create<tensor::EmptyOp>(loc, transpShape, elemType);
+  // If there's a tile with a scalable size, retrieve its size. ATM only 1
+  // scalable tile is allowed.
+  Value scalableSize;
+  for (auto tile : packOp.getMixedTiles()) {
+    if (tile.is<Value>()) {
+      assert(!scalableSize && "Only one scalable size is supported ATM.");
+      scalableSize = cast<Value>(tile);
+      assert(vector::getConstantVscaleMultiplier(scalableSize) &&
+             "This dynamic shape is not a multiple of vscale!");
+    }
+  }
+
+  Value empty =
+      ShapedType::isDynamicShape(cast<ShapedType>(input.getType()).getShape())
+          ? rewriter.create<tensor::EmptyOp>(loc, transpShape, elemType,
+                                             scalableSize)
+          : rewriter.create<tensor::EmptyOp>(loc, transpShape, elemType);
```
You're probably right that this can be simplified, but it's not immediately obvious to me :) Note that for this example:
```mlir
%0 = tensor.pack %arg0 inner_dims_pos = [3, 2] inner_tiles = [8, 32] into %arg1 : tensor<1x1x32x8xf32> -> tensor<1x1x1x1x8x32xf32>
```

`readSizes` and `readShape` are `[1, 1, 32, 8]` and `[32, 8]`, respectively. I need to take another look 😅

hanhanW replied:
oh I see, I misread the code. The unit dims are not inserted to `readShape`. Perhaps we can add `SmallVector<OpFoldResult> transpShape` to l.1190, and insert the `OpFoldResult` values to the vector and use it here.

banach-space replied:
Updated. There's quite a few vectors with shapes/offsets/etc. I've renamed two of them so that we have e.g. `readShapeForExtractSlice` rather than a bit abstract `readShape`. Let me know what you think :)

banach-space commented on mlir/include/mlir/Dialect/Tensor/Utils/Utils.h:
```diff
@@ -14,12 +14,13 @@
 PadOp createPadHighOp(RankedTensorType type, Value source, Value pad,
-                      bool nofold, Location loc, OpBuilder &builder);
+                      bool nofold, Location loc, OpBuilder &builder,
+                      std::optional<Value> dynOutDim = {});
```
For purely pragmatic reasons, I'd like to stick to one dynamic shape for now :) Would extending this in a separate PR be OK?

hanhanW replied:
Yes, it is okay. :)

banach-space replied:
Support for multiple dynamic dims is now included :)

banach-space committed:
fixup! [mlir][tensor] Extend the logic to generalise tensor.pack

* Use `OpFoldResult` for the output shape of tensor::EmptyOp (to
  simplify the invocation)
* Rename `readShape` as `readShapeForExtractSlice`
* Rename `transpShape` as `transShapeForEmpty`

banach-space committed:
fixup! fixup! [mlir][tensor] Extend the logic to generalise tensor.pack

Allow multiple dynamic dims and remove the dependency on #109667

banach-space commented:
Hey @hanhanW , would you be able to take another look? Here's an overview of changes since your last comment:

1. Removed the dependency on #109667 and instead included all the required updates for `createPadHighOp` in this PR. Happy to extract those into a separate PR if that helps (though that will be tricky to test on its own).

2. Relaxed the requirement that _only one_ output dim can be dynamic.

Thanks!

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
```diff
@@ -1178,10 +1212,20 @@ LogicalResult GeneralizeOuterUnitDimsPackOpPattern::matchAndRewrite(
+  applyPermutationToVector<OpFoldResult>(transShapeForEmpty, perm);
 
-  Value empty = rewriter.create<tensor::EmptyOp>(loc, transpShape, elemType);
+  // If there's a tile with a dynamic size, retrieve its size. ATM only 1
+  // dynamic tile is allowed.
+  Value dynDimSize;
+  for (auto tile : packOp.getMixedTiles()) {
+    if (tile.is<Value>()) {
+      assert(!dynDimSize && "Only one scalable size is supported ATM.");
+      dynDimSize = cast<Value>(tile);
+    }
+  }
```
Nit: remove the dead code. We don't need this, right? There are no uses, and the function that should account this is `getPackOpSourceOrPaddedSource`.

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
```diff
@@ -1038,26 +1045,48 @@ static Value getPackOpSourceOrPaddedSource(OpBuilder &builder,
+  // Collect dims for the padded shape.
+  SmallVector<int64_t> paddedShape;
+  for (int64_t dimIdx = 0; dimIdx < inputRank; ++dimIdx) {
+    int64_t inputDimSize = inputType.getDimSize(dimIdx);
```
nit: move it into the if-body because it is only used there.

hanhanW commented on mlir/lib/Dialect/Tensor/Utils/Utils.cpp:
```diff
@@ -16,28 +16,49 @@
+    AffineExpr d0, d1;
+    bindDims(b.getContext(), d0, d1);
+    OpFoldResult sourceDim = tensor::getMixedSize(b, loc, source, idx);
+    Value outDim = isDimDynamic
+                       ? dynOutDims[outDimIdx++]
+                       : b.create<arith::ConstantIndexOp>(loc, val).getResult();
```
I think it is easier if we use OpFoldResult here. You don't need to create the zero constant when it is not needed. The idea is to avoid creating IR if we don't need to. Can you give it a shot?

```cpp
    OpFoldResult outDim = isDimDynamic
                       ? dynOutDims[outDimIdx++]
                       : b.getIndexAttr(val);
```

or something like:

```cpp
    OpFoldResult outDim = b.getIndexAttr(val);;
    if (isDimDynamic)
        outDim = dynOutDims[outDimIdx++];
```

hanhanW commented on mlir/test/Dialect/Linalg/generalize-tensor-pack.mlir:
```diff
@@ -34,6 +36,59 @@ func.func @simple_pad_and_pack(%input: tensor<5x1xf32>, %output: tensor<1x1x8x2x
+// CHECK-SAME:      %[[SRC:[a-zA-Z0-9]+]]: tensor<5x1xf32>,
+// CHECK-SAME:      %[[DEST:[a-zA-Z0-9]+]]: tensor<1x1x?x2xf32>,
+// CHECK-SAME:      %[[PAD_VAL:[a-zA-Z0-9]+]]: f32) -> tensor<1x1x?x2xf32> {
+// CHECK:           %[[C2:.+]] = arith.constant 2 : index
+// CHECK:           %[[C8:.+]] = arith.constant 8 : index
+// CHECK:           %[[VS:.+]] = vector.vscale
```
nit: use CHECK-DAG which is more robust in this case.

hanhanW reviewed (commented):
LGTM, just few nits!

banach-space committed:
fixup! fixup! fixup! [mlir][tensor] Extend the logic to generalise tensor.pack

Incorporating suggestions from @hanhanW

hanhanW reviewed (approved):
Thanks for addressing my comments, LGTM!

banach-space committed:
fixup! fixup! fixup! fixup! [mlir][tensor] Extend the logic to generalise tensor.pack

Remove comment that's no longer valid
